src/Microsoft.AspNetCore.Http/HttpContextAccessor.cs

```diff
@@ -7,21 +7,39 @@ namespace Microsoft.AspNetCore.Http
 {
     public class HttpContextAccessor : IHttpContextAccessor
     {
-        private static AsyncLocal<(string traceIdentifier, HttpContext context)> _httpContextCurrent = new AsyncLocal<(string traceIdentifier, HttpContext context)>();
+        private static AsyncLocal<HttpContextHolder> _httpContextCurrent = new AsyncLocal<HttpContextHolder>();
 
         public HttpContext HttpContext
         {
             get
             {
-                var value = _httpContextCurrent.Value;
-                // Only return the context if the stored request id matches the stored trace identifier
-                // context.TraceIdentifier is cleared by HttpContextFactory.Dispose.
-                return value.traceIdentifier == value.context?.TraceIdentifier ? value.context : null;
+                return  _httpContextCurrent.Value?.Context;
             }
             set
             {
-                _httpContextCurrent.Value = (value?.TraceIdentifier, value);
+                var holder = _httpContextCurrent.Value;
+                if (holder != null)
+                {
+                    // Kill current HttpContext trapped in the AsyncLocals, as its done.
+                    holder.Context = null;
```

Comment:
Clever!

Reply:
If the context starts throwing https://github.com/aspnet/AspNetCore/issues/3586; then it wouldn't necessarily have to clear it here?

The change of holder (below) would still prevent it being shared across requests.

Reply:
People use null to detect if they are in a request aware place.